### PR TITLE
Modbus Controller added some features

### PR DESCRIPTION
**Pull request by 0x3333**
# What does this implement/fix?

Added some helpers and functionalities to Modbus Controller:

* Exposed `get_command_queue_length()` method as public, in my case, it is useful to know if the module has sent all commands(In case we update very frequently).
* If the Modbus module didn't respond, we mark the module `offline`(Currently not exposed to UI or API just via code).
* Added `offline_skip_updates`, if the module is offline, we skip some updates so we don't have to wait too much for the timeout of some module and keep reading others in the same bus.

Default functionally has not changed.

Some background about this PR:

I have a home automation bus with 10 devices. Some devices are high-speed (Mainly input modules) that I update constantly using a custom component(I update every time the command queue is empty, that's why I need `get_command_queue_length()` public). For a high-speed module, if it goes offline, it is marked as faulty and is not updated for `offline_skip_updates` cycles allowing the other modules to be updated without the timeout of the faulty module affecting them too much(My timeout is already low, set to 500ms).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/3163

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
modbus_controller:
  - id: modbus_input_0
    address: 0x1
    modbus_id: modbus_0
    offline_skip_updates: 50
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs). ***Not yet***

### CODE REVIEW

File: esphome/components/modbus_controller/modbus_controller.h [hunk 1]
```diff
@@ -409,7 +409,8 @@ class ModbusCommandItem {
 
 class ModbusController : public PollingComponent, public modbus::ModbusDevice {
  public:
-  ModbusController(uint16_t throttle = 0) : command_throttle_(throttle){};
+  ModbusController(uint16_t throttle = 0, uint16_t offline_skip_updates = 0)
+      : command_throttle_(throttle), offline_skip_updates_(offline_skip_updates){};
```
Comment (jesserockz) [inline esphome/components/modbus_controller/modbus_controller.h]: We dont need the variables in both the constructor and as setters. Lets just change to setters only to match the codebase.
```suggestion
```

File: esphome/components/modbus_controller/__init__.py [hunk 2]
```diff
@@ -206,8 +208,11 @@ async def add_modbus_base_properties(
 
 
 async def to_code(config):
-    var = cg.new_Pvariable(config[CONF_ID], config[CONF_COMMAND_THROTTLE])
-    cg.add(var.set_command_throttle(config[CONF_COMMAND_THROTTLE]))
+    command_throttle = config[CONF_COMMAND_THROTTLE]
+    offline_skip_updates = config[CONF_OFFLINE_SKIP_UPDATES]
+    var = cg.new_Pvariable(config[CONF_ID], command_throttle, offline_skip_updates)
+    cg.add(var.set_command_throttle(command_throttle))
+    cg.add(var.set_offline_skip_updates(offline_skip_updates))
```
Comment (jesserockz) [inline esphome/components/modbus_controller/__init__.py]:
```suggestion
    var = cg.new_Pvariable(config[CONF_ID])
    cg.add(var.set_command_throttle(config[CONF_COMMAND_THROTTLE]))
    cg.add(var.set_offline_skip_updates(config[CONF_OFFLINE_SKIP_UPDATES]))
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/modbus_controller/modbus_controller.h (hunk 1 above)
Comment (0x3333) [reply esphome/components/modbus_controller/modbus_controller.h]: `throttle` was already present in the constructor, should I remove it and let only the setter?

Based on the next comment, sounds like that. Will remove.

**Review by jesserockz — APPROVED** [review]

**Comment by 0x3333** [comment]
I'll create a PR regarding the docs.

**Comment by 0x3333** [comment]
Created Docs [PR](https://github.com/esphome/esphome-docs/pull/3163)

**Pull request merged**